File: app/src/main/java/com/star_zero/example/androidmvvm/domain/Entity.java
package com.star_zero.example.androidmvvm.domain;

import android.databinding.BaseObservable;
import android.support.annotation.NonNull;

public abstract class Entity<T extends Identifier> extends BaseObservable {

    @NonNull
    public final T id;

    protected Entity(@NonNull T id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Entity<?> entity = (Entity<?>) o;

        return id.equals(entity.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }
}

File: app/src/main/kotlin/com/star_zero/example/androidmvvm/domain/Entity.kt
package com.star_zero.example.androidmvvm.domain

import android.databinding.BaseObservable

abstract class Entity<out T : Identifier>(val id: T) : BaseObservable() {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as Entity<*>

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}
